src/compiler/parser.ts

```diff
@@ -6163,6 +6168,7 @@ namespace Parser {
     }
 
     function parseTypeAssertion(): TypeAssertion {
+        Debug.assert(scriptKind === ScriptKind.TS || scriptKind === ScriptKind.TSX, "Type assertions should never be parsed outside of TS/TSX; they should either be comparisons or JSX.");

```

Comment:
On second reading, I this should just be `languageVariant !== LanguageVariant.JSX`? We can't actually parse out a `TypeAssertion` in TSX either.